### PR TITLE
feat!: add initial rate limits

### LINKED ISSUE #3041: add rate limits cf gov proposal

### PR DESCRIPTION
**Pull request by mpoke**
## Description

Closes: https://github.com/cosmos/gaia/issues/3041

Add initial rate limits cf. https://www.mintscan.io/cosmos/proposals/890

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [x] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [x] Added `!` to the type prefix if API, client, or state breaking change (i.e., requires minor or major version bump)
* [x] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/gaia/blob/main/CONTRIBUTING.md#pr-targeting))
* [x] Provided a link to the relevant issue or specification
* [x] Followed the guidelines for [building SDK modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules)
* [x] Included the necessary unit and integration [tests](https://github.com/cosmos/gaia/blob/main/CONTRIBUTING.md#testing)
* [x] Added a changelog entry in `.changelog` (for details, see [contributing guidelines](https://github.com/cosmos/gaia/blob/main/CONTRIBUTING.md#changelog))
* [x] Included comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] Updated the relevant documentation or specification
* [x] Reviewed "Files changed" and left comments if necessary 
* [ ] Confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] confirmed all author checklist items have been addressed 
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage

### CODE REVIEW

**Review by MSalopek — COMMENTED** [review]
Can you run `make lint-fix` to appease the linter?

**Review by MSalopek — APPROVED** [review]

File: app/upgrades/v16/upgrades.go [hunk 1]
```diff
@@ -26,7 +70,76 @@ func CreateUpgradeHandler(
...
 		return vm, err
 	}
 }
+
+// Add rate limits as per https://www.mintscan.io/cosmos/proposals/890
+func AddRateLimits(ctx sdk.Context, k ratelimitkeeper.Keeper) error {
+	ctx.Logger().Info("Adding rate limits...")
+
+	// Osmosis
+	msg := RateLimits["osmosis-1"]
```
Comment (insumity) [inline app/upgrades/v16/upgrades.go:100]: Why don't we have a `for` loop over `RateLimits` to add all rate limits at once? If ordering is of importance, I guess we could make `RateLimits` a `slice` and not a `map`?

File: app/upgrades/v16/upgrades_test.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	ratelimitkeeper := gaiaApp.RatelimitKeeper
+
+	// mint atoms
+	amount := sdk.NewCoin(v16.RateLimitDenom, AtomSupply)
+	amountCoins := sdk.NewCoins(amount)
+	err := gaiaApp.BankKeeper.MintCoins(ctx, minttypes.ModuleName, amountCoins)
+	require.NoError(t, err)
+
+	// mock IBC channels
+	gaiaApp.IBCKeeper.ChannelKeeper.SetChannel(ctx, transfertypes.PortID, v16.RateLimits["osmosis-1"].ChannelId, channeltypes.Channel{})
```
Comment (insumity) [inline app/upgrades/v16/upgrades_test.go:38]: Similarly as above, why not have a `for` loop for this?

File: app/upgrades/v16/upgrades.go [hunk 3]
```diff
@@ -1,15 +1,59 @@
...
 	upgradetypes "github.com/cosmos/cosmos-sdk/x/upgrade/types"
 
 	"github.com/cosmos/gaia/v16/app/keepers"
 )
 
+var RateLimits = map[string]ratelimittypes.MsgAddRateLimit{
+	"osmosis-1": {
+		MaxPercentSend: sdkmath.NewInt(5),
+		MaxPercentRecv: sdkmath.NewInt(5),
+		ChannelId:      "channel-141",
```
Comment (insumity) [inline app/upgrades/v16/upgrades.go:26]: I guess `ChannelId`s are taken from the proposal but how were those initially found?
From what I see [here](https://ping.pub/cosmos/ibc/connection/connection-257) we have multiple channels under `transfer` between the Hub and Osmosis. Might be nice to have a comment on this.

**Review by insumity — APPROVED** [review]

File: app/upgrades/v16/upgrades.go (hunk 1 above)
Comment (mpoke) [reply app/upgrades/v16/upgrades.go:100]: Sure, but given that the code is already written, I don't see any issues with it.

File: app/upgrades/v16/upgrades_test.go (hunk 2 above)
Comment (mpoke) [reply app/upgrades/v16/upgrades_test.go:38]: Same answer as above.

File: app/upgrades/v16/upgrades.go (hunk 3 above)
Comment (mpoke) [reply app/upgrades/v16/upgrades.go:26]: Creating transfer channels is permissionless. This means that there can be multiple channels between two chains. However, social consensus chooses one as the canonical transfer channel. In the case of Osmosis, this is channel-141 -- the channel with all the IBC transfers (see https://www.mintscan.io/cosmos/relayers). Another way to look at this is to check the assets on the Hub https://www.mintscan.io/cosmos/assets. I don't think adding a comment with a link helps as this link might change in the future.

**Comment by MSalopek** [comment]
This does not work as expected. I'm putting it in draft until I figure it out.

This is an actual issue, the testnet would not be able to go through an upgrade.
```shell
➜  gaia git:(marius/3041-rate-limits) ✗ gaiad start --x-crisis-skip-assert-invariants --home ./build/.gaia
1:06PM INF starting node with ABCI Tendermint in-process module=server
1:06PM INF service start impl=multiAppConn module=proxy msg="Starting multiAppConn service"
1:06PM INF service start connection=query impl=localClient module=abci-client msg="Starting localClient service"
1:06PM INF service start connection=snapshot impl=localClient module=abci-client msg="Starting localClient service"
1:06PM INF service start connection=mempool impl=localClient module=abci-client msg="Starting localClient service"
1:06PM INF service start connection=consensus impl=localClient module=abci-client msg="Starting localClient service"
1:06PM INF service start impl=EventBus module=events msg="Starting EventBus service"
1:06PM INF service start impl=PubSub module=pubsub msg="Starting PubSub service"
1:06PM INF service start impl=IndexerService module=txindex msg="Starting IndexerService service"
1:06PM INF ABCI Handshake App Info hash="��h����\x12�MI\x14�\x0e|'kmV\fǘ�k�T\tG\x1b�f�" height=14 module=consensus protocol-version=0 software-version=marius/3041-rate-limits-abe6fefd5f3287ed8e509176e087f9eb2f695ac3
1:06PM INF ABCI Replay Blocks appHeight=14 module=consensus stateHeight=14 storeHeight=15
1:06PM INF Replay last block using real app module=consensus
1:06PM INF applying upgrade "v16" at height: 15 module=server
1:06PM INF Starting module migrations... module=server
1:06PM INF migrating module provider from version 2 to version 3 module=server
1:06PM INF migrating module provider from version 3 to version 4 module=server
1:06PM INF adding a new module: ratelimit module=server
1:06PM INF Adding rate limits... module=server
panic: unable to add rate limit on channel-141 to osmosis-1: channel does not exist

goroutine 10 [running]:
github.com/cosmos/cosmos-sdk/x/upgrade/keeper.Keeper.ApplyUpgrade({{_, _}, _, {_, _}, {_, _}, _, {_, _}, ...}, ...)
	github.com/cosmos/cosmos-sdk@v0.47.11-ics-lsm/x/upgrade/keeper/keeper.go:361 +0x394
github.com/cosmos/cosmos-sdk/x/upgrade.BeginBlocker(_, {{0x1070e2dc0, 0x1087f6580}, {0x1070f8660, 0x140015ffec0}, {{0xb, 0x0}, {0x1400200ea10, 0xb}, 0xf, ...}, ...}, ...)
	github.com/cosmos/cosmos-sdk@v0.47.11-ics-lsm/x/upgrade/abci.go:84 +0x80c
github.com/cosmos/cosmos-sdk/x/upgrade.AppModule.BeginBlock(...)
	github.com/cosmos/cosmos-sdk@v0.47.11-ics-lsm/x/upgrade/module.go:161
github.com/cosmos/cosmos-sdk/types/module.(*Manager).BeginBlock(_, {{0x1070e2dc0, 0x1087f6580}, {0x1070f8660, 0x140015ffec0}, {{0xb, 0x0}, {0x1400200ea10, 0xb}, 0xf, ...}, ...}, ...)
```

**Comment by MSalopek** [comment]
> This does not work as expected. I'm putting it in draft until I figure it out.
> 
> This is an actual issue, the testnet would not be able to go through an upgrade.
> ....

Fixed in https://github.com/cosmos/gaia/pull/3042/commits/075ae6ffb7a6cad7a2987c4d5f996db514c39e4c

**Pull request merged**